### PR TITLE
[cloud_firestore] support for Firestore.Source

### PR DESCRIPTION
**Pull request by Stone2517**
## Description

*This PR supports for retrieving documents from cache or from the server*

## Related Issues

*This PR supports only android not ios. This is my first PR on GitHub and I don't know how to create it sorry for that.*

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [ ] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Comment by Stone2517** [comment]
> Thanks for your pull request. It looks like this may be your first contribution to a Google open source project (if not, look below for help). Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).
> 📝 Please visit https://cla.developers.google.com/ to sign.
> Once you've signed (or fixed any issues), please reply here (e.g. I signed it!) and we'll verify it.
> 
> What to do if you already signed the CLA
> Individual signers
> 
> It's possible we don't have your GitHub username or you're using a different email address on your commit. Check your existing CLA data and verify that your email is set on your git commits.
> 
> Corporate signers
> 
> Your company has a Point of Contact who decides which employees are authorized to participate. Ask your POC to be added to the group of authorized contributors. If you don't know who your Point of Contact is, direct the Google project maintainer to go/cla#troubleshoot (Public version).
> The email used to register you as an authorized contributor must be the email used for the Git commit. Check your existing CLA data and verify that your email is set on your git commits.
> The email used to register you as an authorized contributor must also be attached to your GitHub account.
> 
> ℹ️ Googlers: Go here for more info.
I signed it!

File: packages/cloud_firestore/lib/src/document_reference.dart [hunk 1]
```diff
@@ -71,11 +71,12 @@ class DocumentReference {
   /// Reads the document referenced by this [DocumentReference].
   ///
   /// If no document exists, the read will return null.
-  Future<DocumentSnapshot> get() async {
+  /// Valid options for source are 'server', 'cache' or '' which is default
+  Future<DocumentSnapshot> get({String source = ''}) async {
```
Comment (collinjackson) [inline packages/cloud_firestore/lib/src/document_reference.dart]: I think that `source` should have type `enum FirestoreSource` (which you can define in this file). Although the JavaScript SDK uses a String type it is IMO more idiomatic in Dart to use an enum.

**Comment by Stone2517** [comment]
> Thanks for the contribution. Can you add the iOS implementation and unit tests? I wonder if it's possible to write a good integration test for this one. At the very least we should test that some data comes back from all the possible

**Comment by Stone2517** [comment]
> Thanks for the contribution. Can you add the iOS implementation and unit tests? I wonder if it's possible to write a good integration test for this one. At the very least we should test that some data comes back from all the possible sources.

Thank you for the support I will try to add ios implementation and unit tests. I'm new here.

File: packages/cloud_firestore/lib/src/document_reference.dart (hunk 1 above)
Comment (Stone2517) [reply packages/cloud_firestore/lib/src/document_reference.dart]: Hey @collinjackson I have added enum FirestoreSource to the source.

**Comment by cbenhagen** [comment]
@Stone2517 why did you close this PR?

**Comment by Stone2517** [comment]
@cbenhagen I don't have Mac to run and test it. so I closed it. Its easy to implement on ios. 
You have to manually do it.
Change the code on 452 line DocumentReference#get https://github.com/flutter/plugins/blob/master/packages/cloud_firestore/ios/Classes/CloudFirestorePlugin.m

`[document getDocumentWithCompletion:^(FIRDocumentSnapshot *_Nullable snapshot, NSError *_Nullable error) {`

to 
`[document getDocumentWithSource:FIRFirestoreSourceCache
                   completion:^(FIRDocumentSnapshot *_Nullable snapshot, NSError *error) {`

on line 521 
` [query getDocumentsWithCompletion:^(FIRQuerySnapshot *_Nullable snapshot, NSError *_Nullable error) {`
to
` [query getDocumentsWithSource: FIRFirestoreSourceCache
              completion:^(FIRQuerySnapshot *_Nullable snapshot,NSError *_Nullable error) {`

If you want to get data from server or default pass the source and filter the arguments and replace FIRFirestoreSourceDefault with  FIRFirestoreSourceServer or FIRFirestoreSourceCache.
Change it As I did in CloudFirestorePlugin.java.

**Pull request closed without merging**